certora/diff/access_manager_AccessManager.sol.patch

```diff
@@ -0,0 +1,61 @@
+--- access/manager/AccessManager.sol	2023-09-18 16:59:06.975566974 +0200
++++ access/manager/AccessManager.sol	2023-09-18 17:32:34.590396666 +0200
+@@ -6,7 +6,6 @@
+ import {IAccessManaged} from "./IAccessManaged.sol";
+ import {Address} from "../../utils/Address.sol";
+ import {Context} from "../../utils/Context.sol";
+-import {Multicall} from "../../utils/Multicall.sol";
+ import {Math} from "../../utils/math/Math.sol";
+ import {Time} from "../../utils/types/Time.sol";
+ 
+@@ -48,7 +47,7 @@
+  * mindful of the danger associated with functions such as {{Ownable-renounceOwnership}} or
+  * {{AccessControl-renounceRole}}.
+  */
+-contract AccessManager is Context, Multicall, IAccessManager {
++contract AccessManager is Context, IAccessManager {
```

Comment:
AFAIK, certora is not smart enough to understand that the target of the delegate call in multicall is `address(this)`. This means that certora believes anything can change (in storage) through multicall.

A consequence is that basically any invariant would get broken by multicall.

So we decide to test without it, as we know adding it latter would not change the outcome of the verification.

Reply:
> This means that certora believes anything can change (in storage) through multicall.

This is pretty bad imo. How can we know for sure that adding it later wouldn't change the verification in the future? I was thinking that some flow through Multicall may mess up with the state (e.g. chaining schedule + execution in a weird way).

This shouldn't be an immediate concern because currently we're sure Multicall doesn't change the verification, but if we make an update in the future (eg. executors), we may want to revisit the Multicall in the Formal Verification spec.

Perhaps adding a comment about this to the patch would be at least helpful when adding features to the contract.

Reply:
```
* NOTE: The FV version of this contract doesn't include Multicall because CVL HAVOCs on any `delegatecall`.
* Addition of new features to this contract should be added considering how they can be composed through Multicall.
```

Reply:
There is something about multicall I discused with Fran multiple time, and that maybe we should formally document. I'm taking this opportunity to write it down, maybe it will answer some of your concerns.

Multicall allows someone to call multiple function in a batch. Its particularly usefull for EOA, because smart wallets can already do that. Also, there is really no execution difference between doing a multicall and doing multiple transaction that end up in the same block (with nothing in between).

In particular, if we have a rule that is value X (in storage) can only change if you call function F, you should not be worried about multicall and interractions between function:

Lets assume you have function F1 and F2, and now assume that the value X changes when doing F1 + F2 in a multicall. Either the value changed during the first F1 call, or it changed during the second F2 call (or it change twice). In both cases, if you extract the F1 call and the F2 call, one of the two must change the value alone. That should be detected by FV. 

Said otherwize, if F1 does not change X given an arbitrary initial state), and if F2 does not change X (given an arbitrary initial state), then F1 + F2 is not going to change X.

This means that if something can happen using multicall, it can already happen without multicall, and FV should detect it with multicall disabled.

Reply:
Thanks for the explanation! I see how Multicall should not modify the guarantees that are already in the contract.

Still, let's partially keep the note because I still see contextual value in it:

```
* NOTE: The FV version of this contract doesn't include Multicall because CVL HAVOCs on any `delegatecall`.
```